File: app-laravel/app/Http/Controllers/WebhookController.php
<?php

declare(strict_types=1);

namespace App\Http\Controllers;

use App\Http\Resources\ApiResponse;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Support\Str;
use Src\Shared\Infrastructure\Webhook\WebhookModel;

/**
 * WEBHOOK CONTROLLER
 * ==================
 * İstifadəçilərə webhook yaratmaq, silmək, siyahılamaq imkanı verir.
 *
 * WEBHOOK CRUD:
 * POST   /api/webhooks          → Yeni webhook yarat
 * GET    /api/webhooks          → Webhook siyahısı
 * DELETE /api/webhooks/{id}     → Webhook sil
 * PATCH  /api/webhooks/{id}     → Webhook aktivləşdir/deaktiv et
 */
class WebhookController extends Controller
{
    /**
     * POST /api/webhooks
     * Yeni webhook yarat.
     *
     * Request body:
     * {
     *   "url": "https://my-system.com/webhooks/callback",
     *   "events": ["order.created", "payment.completed", "payment.failed"]
     * }
     *
     * secret_key avtomatik yaradılır — istifadəçi bunu bir dəfə görür.
     * Bu key ilə HMAC imzanı yoxlaya bilər.
     */
    public function store(Request $request): JsonResponse
    {
        $validated = $request->validate([
            'url' => ['required', 'url', 'max:500'],
            'events' => ['required', 'array', 'min:1'],
            'events.*' => ['string', 'in:order.created,order.cancelled,order.paid,payment.completed,payment.failed,product.low_stock'],
        ]);

        $secretKey = Str::random(40);

        $webhook = WebhookModel::create([
            'user_id' => $request->user()->id,
            'url' => $validated['url'],
            'events' => $validated['events'],
            'secret_key' => $secretKey,
            'is_active' => true,
        ]);

        return ApiResponse::success(
            data: [
                'webhook_id' => $webhook->id,
                'secret_key' => $secretKey, // Yalnız BİR DƏFƏ göstərilir!
            ],
            message: 'Webhook yaradıldı. Secret key-i qeyd edin — bir daha göstərilməyəcək!',
            code: 201,
        );
    }

    /**
     * GET /api/webhooks
     * İstifadəçinin webhook-ları.
     */
    public function index(Request $request): JsonResponse
    {
        $webhooks = WebhookModel::where('user_id', $request->user()->id)
            ->select(['id', 'url', 'events', 'is_active', 'created_at'])
            ->get();

        return ApiResponse::success(data: $webhooks);
    }

    /**
     * DELETE /api/webhooks/{id}
     */
    public function destroy(Request $request, string $id): JsonResponse
    {
        $webhook = WebhookModel::where('id', $id)
            ->where('user_id', $request->user()->id)
            ->firstOrFail();

        $webhook->delete();

        return ApiResponse::success(message: 'Webhook silindi');
    }

    /**
     * PATCH /api/webhooks/{id}
     * Aktivləşdir/deaktiv et.
     */
    public function toggle(Request $request, string $id): JsonResponse
    {
        $webhook = WebhookModel::where('id', $id)
            ->where('user_id', $request->user()->id)
            ->firstOrFail();

        $webhook->update(['is_active' => !$webhook->is_active]);

        return ApiResponse::success(
            data: ['is_active' => $webhook->is_active],
            message: $webhook->is_active ? 'Webhook aktivləşdirildi' : 'Webhook deaktiv edildi',
        );
    }
}

File: app/golang/internal/webhook/webhook_controller.go
// Package webhook — Webhook CRUD controller
package webhook

import (
	"encoding/json"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"github.com/orkhan/ecommerce/internal/middleware"
	"github.com/orkhan/ecommerce/internal/shared/infrastructure/api"
	webhookInfra "github.com/orkhan/ecommerce/internal/shared/infrastructure/webhook"
	"gorm.io/gorm"
)

// Controller — webhook idarəetmə endpoint-ləri
type Controller struct {
	db *gorm.DB
}

func New(db *gorm.DB) *Controller { return &Controller{db: db} }

// POST /api/webhooks  body: {url, events: ["order.created", ...]}
func (c *Controller) Store(ctx *gin.Context) {
	uid := ctx.MustGet(middleware.UserIDKey).(uuid.UUID)
	var body struct {
		URL    string   `json:"url" binding:"required,url"`
		Events []string `json:"events" binding:"required,min=1"`
	}
	if err := ctx.ShouldBindJSON(&body); err != nil {
		ctx.Error(err)
		return
	}

	eventsJSON, _ := json.Marshal(body.Events)
	w := webhookInfra.Webhook{
		ID:       uuid.New(),
		UserID:   uid,
		URL:      body.URL,
		Secret:   uuid.New().String(),
		Events:   string(eventsJSON),
		IsActive: true,
	}
	if err := c.db.Create(&w).Error; err != nil {
		ctx.Error(err)
		return
	}
	ctx.JSON(http.StatusCreated, api.SuccessWithMessage(w, "Webhook yaradıldı"))
}

// GET /api/webhooks
func (c *Controller) Index(ctx *gin.Context) {
	uid := ctx.MustGet(middleware.UserIDKey).(uuid.UUID)
	var hooks []webhookInfra.Webhook
	if err := c.db.Where("user_id = ?", uid).Find(&hooks).Error; err != nil {
		ctx.Error(err)
		return
	}
	ctx.JSON(http.StatusOK, api.Success(hooks))
}

// PATCH /api/webhooks/:id  → toggle
func (c *Controller) Toggle(ctx *gin.Context) {
	id, err := uuid.Parse(ctx.Param("id"))
	if err != nil {
		ctx.Error(err)
		return
	}
	var w webhookInfra.Webhook
	if err := c.db.First(&w, "id = ?", id).Error; err != nil {
		ctx.Error(err)
		return
	}
	w.IsActive = !w.IsActive
	if err := c.db.Save(&w).Error; err != nil {
		ctx.Error(err)
		return
	}
	ctx.JSON(http.StatusOK, api.Success(w))
}

// DELETE /api/webhooks/:id
func (c *Controller) Destroy(ctx *gin.Context) {
	id, err := uuid.Parse(ctx.Param("id"))
	if err != nil {
		ctx.Error(err)
		return
	}
	if err := c.db.Delete(&webhookInfra.Webhook{}, "id = ?", id).Error; err != nil {
		ctx.Error(err)
		return
	}
	ctx.JSON(http.StatusOK, api.SuccessWithMessage(nil, "Silindi"))
}

func (c *Controller) RegisterRoutes(authed *gin.RouterGroup) {
	authed.POST("/webhooks", c.Store)
	authed.GET("/webhooks", c.Index)
	authed.PATCH("/webhooks/:id", c.Toggle)
	authed.DELETE("/webhooks/:id", c.Destroy)
}
